File: tests/mv-2c/hosts.pl
#! /usr/bin/perl
use strict;
use warnings;
no warnings "uninitialized";
use re::engine::RE2;
use JSON::JsonModel;
use constant JMC_VERSION => '2';


sub json_model_1($$$);
my %check_model_map;

sub json_model_1($$$)
{
    my ($val, $path, $rep) = @_;
    if (! jm_is_object($val))
    {
        return 0;
    }
    my $res;
    scalar keys %$val;
    while (my ($prop, $pval) = each %$val)
    {
        if ($prop eq 'host')
        {
            $res = jm_is_string($pval) && jm_is_host($pval, undef, undef) && length $pval <= 255;
            if (! $res)
            {
                return 0;
            }
            next;
        }
        elsif ($prop eq 'ipv4')
        {
            $res = jm_is_string($pval) && jm_is_ip4($pval, undef, undef);
            if (! $res)
            {
                return 0;
            }
            next;
        }
        elsif ($prop eq 'ipv6')
        {
            $res = jm_is_string($pval) && jm_is_ip6($pval, undef, undef);
            if (! $res)
            {
                return 0;
            }
            next;
        }
        return 0;
    }
    return 1;
}


sub jm_is_host($$$)
{
    my ($val, $path, $rep) = @_;
    my $res = $val =~ /^([a-z0-9][-a-z0-9]{0,62})(\.([a-z0-9][-a-z0-9]{0,62}))*$/i;
    return $res;
}

sub jm_is_ip4($$$)
{
    my ($val, $path, $rep) = @_;
    my $res = $val =~ /^(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])$/;
    return $res;
}

sub jm_is_ip6($$$)
{
    my ($val, $path, $rep) = @_;
    my $res = $val =~ /^(([0-9a-f]{1,4}:){7}[0-9a-f]{1,4}|([0-9a-f]{1,4}:){1,7}:|([0-9a-f]{1,4}:){1,6}(:[0-9a-f]{1,4}){1}|([0-9a-f]{1,4}:){1,5}(:[0-9a-f]{1,4}){1,2}|([0-9a-f]{1,4}:){1,4}(:[0-9a-f]{1,4}){1,3}|([0-9a-f]{1,4}:){1,3}(:[0-9a-f]{1,4}){1,4}|([0-9a-f]{1,4}:){1,2}(:[0-9a-f]{1,4}){1,5}|[0-9a-f]{1,4}:(:[0-9a-f]{1,4}){1,6}|:(:[0-9a-f]{1,4}){1,7}|::)$/i;
    return $res;
}

# initialization of global variables

our $initialized = 0;

sub check_model_init()
{
    if (!$initialized)
    {
        $initialized = 1;
        %check_model_map = (
            '' => \&json_model_1,
        );
    }
}

sub check_model_free()
{
    if ($initialized)
    {
        $initialized = 0;
        %check_model_map = ();
    }
}

sub check_model_mapper($)
{
    my ($name) = @_;
    die "unexpected model name \"$name\"" unless exists $check_model_map{$name};
    return $check_model_map{$name};
}

sub check_model($$$)
{
    my ($json, $name, $rep) = @_;
    die "unexpected model name \"$name\"" unless exists $check_model_map{$name};
    my $path = defined $rep ? [] : undef;
    return $check_model_map{$name}($json, $path, $rep);
}

#
# SCRIPT EXECUTION
#
check_model_init();
jm_main(\&check_model, \%check_model_map, JMC_VERSION);
check_model_free();

File: tests/mv-2c/hosts.js
#! /bin/env node
// we may need require to load re2
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);
import * as runtime from "json_model_runtime"
const JSON_MODEL_VERSION = "2";

export var check_model_map = new Map()
const jm_is_eth_re = new runtime.RX("^([0-9a-f]{2}:){5}[0-9a-f]{2}$", "i")
const jm_is_host_re = new runtime.RX("^([a-z0-9][-a-z0-9]{0,62})(\\.([a-z0-9][-a-z0-9]{0,62}))*$", "i")
const jm_is_ip4_re = new runtime.RX("^(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])$", "")
const jm_is_ip6_re = new runtime.RX("^(([0-9a-f]{1,4}:){7}[0-9a-f]{1,4}|([0-9a-f]{1,4}:){1,7}:|([0-9a-f]{1,4}:){1,6}(:[0-9a-f]{1,4}){1}|([0-9a-f]{1,4}:){1,5}(:[0-9a-f]{1,4}){1,2}|([0-9a-f]{1,4}:){1,4}(:[0-9a-f]{1,4}){1,3}|([0-9a-f]{1,4}:){1,3}(:[0-9a-f]{1,4}){1,4}|([0-9a-f]{1,4}:){1,2}(:[0-9a-f]{1,4}){1,5}|[0-9a-f]{1,4}:(:[0-9a-f]{1,4}){1,6}|:(:[0-9a-f]{1,4}){1,7}|::)$", "i")

function json_model_1(val, path, rep)
{
    if (! (Object.prototype.toString.call(val) === '[object Object]'))
    {
        return false;
    }
    let res;
    for (const [prop, pval] of Object.entries(val))
    {
        if (prop == "eth")
        {
            res = ((typeof pval === 'string' || pval instanceof String)) && jm_is_eth(pval, null, null);
            if (! res)
            {
                return false;
            }
            continue;
        }
        else if (prop == "host")
        {
            res = ((typeof pval === 'string' || pval instanceof String)) && jm_is_host(pval, null, null) && pval.length <= 255;
            if (! res)
            {
                return false;
            }
            continue;
        }
        else if (prop == "ipv4")
        {
            res = ((typeof pval === 'string' || pval instanceof String)) && jm_is_ip4(pval, null, null);
            if (! res)
            {
                return false;
            }
            continue;
        }
        else if (prop == "ipv6")
        {
            res = ((typeof pval === 'string' || pval instanceof String)) && jm_is_ip6(pval, null, null);
            if (! res)
            {
                return false;
            }
            continue;
        }
        return false;
    }
    return true;
}


const jm_is_eth = (s) => jm_is_eth_re.exec(s) !== null

const jm_is_host = (s) => jm_is_host_re.exec(s) !== null

const jm_is_ip4 = (s) => jm_is_ip4_re.exec(s) !== null

const jm_is_ip6 = (s) => jm_is_ip6_re.exec(s) !== null

var initialized = false

// differed module initializations
export function check_model_init()
{
    if (! initialized)
    {
        initialized = true;
        runtime.jm_set_rx(RegExp)
        check_model_map.set("", json_model_1)
    }
}

// differed module cleanup
export function check_model_free()
{
    if (initialized)
    {
        initialized = false;
    }
}

export function check_model(val, name, rep)
{
    let checker = check_model_map.get(name)
    if (checker === undefined)
        throw `no checker for "${name}"`
    let path = rep !== null ? [] : null

    return checker(val, path, rep)
}

// possibly run as main based on a guess
import main from "json_model_runtime/main.js"

if (import.meta.url.endsWith(process.argv[1]))
    main(check_model_init, check_model, check_model_free)
